src/main/java/us/ihmc/commons/thread/RepeatingTaskThread.java

```diff
@@ -0,0 +1,448 @@
+
+/**
+ * A thread that repeats the execution of a single task. It can do this {@code n} times, continuously, or at a limited rate.
+ * <p>
+ * Unlike {@link Thread}, this class ensures that exceptions are caught and handled.
+ * <p>
+ *  The task to execute may be specified in two ways:
+ *  <ol>
+ *     <li> Providing a {@link RunnableThatThrows} during construction.
+ *     <li> {@code @Override}ing the {@link #runTask()} method.
+ *  </ol>
+ * <p>
+ * Upon construction, this thread will have zero scheduled task executions and the thread will not be started.
+ * Execution can be started in two ways:
+ *  <ol>
+ *     <li> Call {@link #startRepeating()}
+ *     <li> Call {@link #start()}, and {@link #setScheduled(long n)} or {@link #addScheduled(long n)} in any order.
+ *  </ol>
+ * <p>
+ * The execution frequency may be limited with {@link #setFrequencyLimit(double)} and removed using {@link #removeFrequencyLimit()}.
+ * The limit can be modified or removed at any time.
+ * Note that this is only an upper limit. If the task overruns the corresponding period, the task will simply execute again
+ * once its finished, resulting in a potentially jittery, slower frequency.
+ * <p>
+ * To support clear and concise initialization, {@link #setFrequencyLimit} returns {@code this}.
+ * <p>
+ * To allow the currently executing task to finish, if one is executing, and allow the thread to die,
+ * use {@link #kill()} or {@link #blockingKill()}.
+ */
+public class RepeatingTaskThread extends Thread
+{
+   public static final int REPEAT_INDEFINITELY = -1;
+   public static final double UNLIMITED_FREQUENCY = -1.0;
+
+   private final RunnableThatThrows task;
+   private final ExceptionHandler exceptionHandler;
+
+   /**
+    * Becomes {@code true} when the thread is started, and {@code false} when the thread is killed.
+    * Once {@code false}, the task loop will allow the currently executing task (if any) to complete,
+    * and the task loop is exited, allowing the thread to die.
+    */
+   private volatile boolean running = false;
+   
+   /** Execution state of this thread */
+   private final ExecutionState executionState = new ExecutionState();
+
+   /** Throttler used to limit the execution frequency. */
+   private final Throttler throttler = new Throttler();
+
+   /** The period of the set frequency limit. A negative value indicates no limit. */
+   private volatile double periodLowerLimit = UNLIMITED_FREQUENCY;
+
+   public RepeatingTaskThread(String name)
+   {
+      this(DefaultExceptionHandler.MESSAGE_AND_STACKTRACE, name);
+   }
+
+   public RepeatingTaskThread(ExceptionHandler exceptionHandler, String name)
+   {
+      this(null, exceptionHandler, name);
+   }
+
+   public RepeatingTaskThread(RunnableThatThrows task, String name)
+   {
+      this(task, DefaultExceptionHandler.MESSAGE_AND_STACKTRACE, name);
+   }
+
+   public RepeatingTaskThread(RunnableThatThrows task, ExceptionHandler exceptionHandler, String name)
```

Comment:
`name` should come first in the parameters

Reply:
Dang, I was trying to keep the constructor similar to Thread, where name is after the runnable. Not a big deal though, and I also think name being first looks better in code haha

Reply:
In that case the task should always be last, but I did like the way Tomasz had it here

Reply:
Yes, I would prefer if `name` was first and `task` was last

Reply:
I like having exception handler after runnable because you can do something like:
```
RepeatingTaskThread myThread = new RepeatingTaskThread("name", () -> 
{
   // Do stuff
}, exception -> 
{
   // Handle exceptions
});
```
It almost looks like a try catch block, which is very familiar to us. If exception handler came first it'd be a catch try block? :shrug:

Reply:
Hmm, I suppose that's a decent reason to keep the current ordering